taskcluster/client.py

```diff
@@ -219,62 +222,125 @@ def genBewit():
             raise exceptions.TaskclusterFailure('Did not receive a bewit')
 
         u = urllib.parse.urlparse(requestUrl)
+
+        qs = u.query
+        if qs:
+            qs += '&'
+        qs += 'bewit=%s' % bewit
+
         return urllib.parse.urlunparse((
             u.scheme,
             u.netloc,
             u.path,
             u.params,
-            u.query + 'bewit=%s' % bewit,
+            qs,
             u.fragment,
         ))
 
     def _makeApiCall(self, entry, *args, **kwargs):
         """ This function is used to dispatch calls to other functions
         for a given API Reference entry"""
 
-        payload = None
-        _args = list(args)
-        _kwargs = copy.deepcopy(kwargs)
-
-        if 'input' in entry:
-            if len(args) > 0:
-                payload = _args.pop()
-            else:
-                raise exceptions.TaskclusterFailure('Payload is required as last positional arg')
-        apiArgs = self._processArgs(entry, *_args, **_kwargs)
-        route = self._subArgsInRoute(entry, apiArgs)
+        routeParams, payload, query = self._processArgs(entry, *args, **kwargs)
+        route = self._subArgsInRoute(entry, routeParams)
+        if query:
+            route += '?' + urllib.parse.urlencode(query)
         log.debug('Route is: %s', route)
 
         return self._makeHttpRequest(entry['method'], route, payload)
 
-    def _processArgs(self, entry, *args, **kwargs):
-        """ Take the list of required arguments, positional arguments
-        and keyword arguments and return a dictionary which maps the
-        value of the given arguments to the required parameters.
-
-        Keyword arguments will overwrite positional arguments.
+    def _processArgs(self, entry, *_args, **_kwargs):
+        """ Figure out, given an entry, positional and keyword arguments, what
+        the query-string options, payload and api arguments are.
         """
 
+        # We need the args to be a list so we can mutate them
+        args = list(_args)
+        kwargs = copy.deepcopy(_kwargs)
+        log.debug(args)
```

Comment:
If these are being left in, perhaps they should be something like  "${entry} called with args ${args} and kwargs ${kwargs}". Not that it has to be exactly that, but it might be a bit nicer.

It is also possible that I don't remember how python logging works and this is just fine as it is.

Reply:
they're not supposed to be left in, but I couldn't for the life of me remember where i'd typed them.  i'll remove them